nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PublishSlack.java

```diff
@@ -0,0 +1,513 @@
+        .description("Specifies whether or not the contents of the FlowFile should be uploaded as an attachment to the Slack message.")
+        .allowableValues("true", "false")
+        .required(true)
+        .dependsOn(PUBLISH_STRATEGY, PUBLISH_STRATEGY_USE_PROPERTY)
+        .defaultValue("false")
+        .build();
+
+    protected PropertyDescriptor MAX_FILE_SIZE = new PropertyDescriptor.Builder()
+        .name("Max FlowFile Size")
+        .description("The maximum size of a FlowFile that can be sent to Slack. If any FlowFile exceeds this size, it will be routed to failure. " +
+                     "This plays an important role because the entire contents of the file must be loaded into NiFi's heap in order to send the data " +
+                     "to Slack.")
+        .required(true)
+        .dependsOn(SEND_CONTENT_AS_ATTACHMENT, "true")
+        .addValidator(StandardValidators.DATA_SIZE_VALIDATOR)
+        .defaultValue("1 MB")
+        .build();
+
+    protected PropertyDescriptor THREAD_TS = new PropertyDescriptor.Builder()
+        .name("Thread Timestamp")
+        .description("The Timestamp identifier for the thread that this message is to be a part of. If not specified, the message will be a top-level message instead of " +
+                     "being in a thread.")
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+        .required(false)
+        .build();
+
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+        .name("success")
+        .description("FlowFiles are routed to success after being successfully sent to Slack")
+        .build();
+
+    public static final Relationship REL_RATE_LIMITED = new Relationship.Builder()
+        .name("rate limited")
+        .description("FlowFiles are routed to 'rate limited' if the Rate Limit has been exceeded")
+        .build();
+
+    public static final Relationship REL_TOO_LARGE = new Relationship.Builder()
+        .name("too large")
+        .description("FlowFiles are routed to 'too large' if they exceed Slack's limits or the Processor's configured size limit")
+        .build();
+
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+        .name("failure")
+        .description("FlowFiles are routed to 'failure' if unable to be sent to Slack for any other reason")
+        .build();
+
+    private final RateLimit rateLimit = new RateLimit(getLogger());
+
+    private volatile ChannelMapper channelMapper;
+    private volatile App slackApp;
+    private volatile MethodsClient client;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return List.of(ACCESS_TOKEN,
+            CHANNEL,
+            PUBLISH_STRATEGY,
+            MESSAGE_TEXT,
+            CHARACTER_SET,
+            SEND_CONTENT_AS_ATTACHMENT,
+            MAX_FILE_SIZE,
+            THREAD_TS);
```

Comment:
Is there a reason for declaring this in the method as opposed to a private static member variable? That would avoid constructing the list on each method invocation.

Reply:
Nope, will change that.